### ISSUE TITLE
ExternalSignal Azure Function Task waiting bug

### ISSUE THREAD

**Issue by drub0y** [microsoft/AuthJanitor#55]
# TL;DR 
Use `Task.Run` instead of creating and manually starting a `Task` in the `ExternalSignal` Azure Function.

----

# The problem
The `ExternalSignal` Azure Function contains the following code:

https://github.com/microsoft/AuthJanitor/blob/35dfb549ec8a4bd33496c422d4dd5410b287899b/AuthJanitor.Automation.Agent/ExternalSignal.cs#L77-L102

This will not work as intended. The `Task` that is created will seem to always finish almost immediately and the `.Wait(...)` will always return `true` the instant it does.

# TPL history lesson and modern async best practices
As a quick aside, it's rare that you'd want to use the pattern of `new Task(...) + task.Start()`... even back in the "pure" TPL days. It's usually preferable to start tasks via `Task.Factory.StartNew(...)`. [Stephen Toub has a blog post on this topic here](https://devblogs.microsoft.com/pfxteam/task-factory-startnew-vs-new-task-start/). 

Ok, now, that's technically a [historical document](https://i.imgflip.com/wjurk.jpg) and there's actually a bigger problem with either of those approaches in the new world of `async`/`await`. The real problem with the code stems from passing an `async` lambda to the constructor of the `Task` (same would apply to using `StartNew(...)`. [For an explanation on why exactly I'll refer to another great blog post, this time from Stephen Cleary](https://blog.stephencleary.com/2013/08/startnew-is-dangerous.html). From that post specifically, this is the problem that this code has:

> Does not understand async delegates. This is actually the same as point 1 in the reasons why you would want to use StartNew. The problem is that when you pass an async delegate to StartNew, it’s natural to assume that the returned task represents that delegate. However, since StartNew does not understand async delegates, what that task actually represents is just the beginning of that delegate. This is one of the first pitfalls that coders encounter when using StartNew in async code.

So, in this this code's specific case what's going to happen is that the `rekeyingTask` is going to start and as soon as the first `await` statement is reached that triggers an actual async call it's returning and that's going to cause the `Task` to complete. This is a problem because the logic right afterwards is to wait for the `rekeyingTask` for a certain amount of time, but the task will be considered completed almost instantly while the actual work is still off running asynchronously.

# LongRunning...? Not really.
Now, I assume the reason this code was written this way in the first place was because someone thought they should shuffle the execution off to a dedicated background thread and that's why `TaskCreationOptions.LongRunning` is specified. However, this actually makes no sense because the work the task does itself is not compute bound. The task itself executes `async` methods itself which means it is completely safe to run on the task pool because it is not going to "hog" the task pool thread for any significant amount of time.

# The modern async friendly solution
The solution is to prefer using `Task.Run` for these kinds of workloads. `Task.Run` is the modern way to "safely" enqueue `async` work to the task pool. Easy peasy, done. Almost...

# Waiting without blocking
We have one more problem. The code today uses `.Wait(...)` to detect a timeout. `Wait(...)` is a blocking call and will block the thread that the Azure Function runtime used to invoke the `ExternalSignal::Run` method. This is... not good, right? So how can we solve this? Easy, use a `Task.Delay(...)` to represent the timeout and have the `Run` method await a `Task::WhenAny`  of the `rekeyingTask` and the timeout task instead. Doing this is the last step to making the `Run` method fully asynchronous.
